### PR TITLE
Add dictionary to search

### PR DESCRIPTION
**Pull request by blake-sc**

### CODE REVIEW

File: server/server/search/dictionaries.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+from elasticsearch import ConnectionError
+from flask import request, current_app
+from flask_restful import Resource
+
+from search import query as query_search
+from collections import Counter
+
+
+from common.arangodb import get_db
+
```
Comment (gradam) [inline server/server/search/dictionaries.py]: There should be 2 blank lines

File: server/server/search/dictionaries.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import json
```
Comment (gradam) [inline server/server/search/dictionaries.py]: redundant import

File: server/server/search/dictionaries.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+import lxml.html
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+from elasticsearch import ConnectionError
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+from flask import request, current_app
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+from flask_restful import Resource
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+from search import query as query_search
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+from collections import Counter
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+def cleanly_truncate_html(html_string, length):
+    # TODO: The function name is aspirational, at the moment it crudely
+    # butchers the HTML. Need to discuss further before choosing implementation.
+    content = html_string
+    if len(content) >= length:
+        return content[:length] + '…'
+    else:
+        return content
+
```
Comment (gradam) [inline server/server/search/dictionaries.py]: There should be 2 blank lines

File: server/server/search/dictionaries.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    
+def search(word, truncate_length=1000):
+    db = get_db()
+    
+    results = list(db.aql.execute('''
+        FOR doc IN dictionary_full
+            FILTER doc.word == @word OR doc.word_ascii == @word
+            RETURN KEEP(doc, "dictname", "lang_to", "lang_from", "word", "word_ascii", "text")
+    ''',
+    bind_vars={'word': word}))
```
Comment (gradam) [inline server/server/search/dictionaries.py]: continuation line under-indented for visual indent
https://www.python.org/dev/peps/pep-0008/#indentation

File: server/server/search/dictionaries.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        results = exact_results
+    else:
+        # We could have multiple inexact results i.e. pali would match both
+        # pāli and palī. We'll just use the word with the most
+        # written about it.
+        word_to_use = sorted(results, key=lambda r: len(r['text']))[-1]['word']
+        results = [r for r in results if r['word'] == word_to_use]
+    
+    # We can have multiple hit: one per dictionary. We want to summarize them.
+    gloss = None
```
Comment (gradam) [inline server/server/search/dictionaries.py]: Variable not used

File: server/server/search/dictionaries.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            "title": ""
+        },
+        "highlight": {
+            "content": [content]
+        },
+        "url": f"/define/{word}",
+        "category": "dictionary",
+    }
+        
+
```
Comment (gradam) [inline server/server/search/dictionaries.py]: Redundant blank lines

File: server/server/search/dictionaries.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+            best_result_found = True
+        
+        if not best_result_found:
+            if not best_result or (len(result['text']) > len(best_result['text'])):
+                best_result = result
+    
+    content = cleanly_truncate_html(best_result['text'], truncate_length)
+    
+    return {
+        "heading": {
```
Comment (gradam) [inline server/server/search/dictionaries.py]: Please use single quotes as it is the convention in our code base.

File: server/server/search/dictionaries.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+            "division": best_result['dictname'],
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+            "subhead": "",
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+            "title": ""
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        },
+        "highlight": {
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+            "content": [content]
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+        },
+        "url": f"/define/{word}",
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+        "category": "dictionary",
```
Comment (gradam) [inline server/server/search/dictionaries.py]: ditto

File: server/server/search/dictionaries.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+    word = results[0]['word']
+    best_result = None
+    best_result_found = False
+    for result in results:
+        if result['dictname'] == 'ncped':
+            best_result = result
+            best_result_found = True
+        
+        if not best_result_found:
+            if not best_result or (len(result['text']) > len(best_result['text'])):
```
Comment (gradam) [inline server/server/search/dictionaries.py]: `not best_result` always evaluate to `True` here. So no need for that.

File: server/server/search/view.py [hunk 22]
```diff
@@ -77,7 +78,8 @@ def get(self):
 
         if query is None:
             return json.dumps({'error': '\'query\' param is required'}), 422
-
+            
+        results = {'total': 0, 'hits': []}
```
Comment (gradam) [inline server/server/search/view.py]: Variable not used

File: server/server/search/dictionaries.py (hunk 21 above)
Comment (blake-sc) [reply server/server/search/dictionaries.py]: best_result can still be none if the first result is not ncped but I've rewritten that section to just use a break

File: server/server/search/view.py (hunk 22 above)
Comment (blake-sc) [reply server/server/search/view.py]: Variable used now. Elasticsearch is a good deal more likely to fail to return results than ArangoDB so I'm writing the code with the possibility that we'll make it so it still returns the dictionary card if elasticsearch times out or somesuch.

**Pull request merged**